File: partition_list.py
class ListNode:
    def __init__(self,x):
        self.val=x
        self.next=None
class Solution:
    def partition(self,head,x):
        dummy1=ListNode(-1)
        dummy2=ListNode(-1)
        phead1=dummy1
        phead2=dummy2

        while head:
            if head.val<x:
                phead1.next=head
                phead1=phead1.next
            else:
                phead2.next=head
                phead2=phead2.next
            head=head.next
        phead2.next=None
        phead1.next=dummy2.next
        return dummy1.next
        
                

        
        


if __name__=='__main__':
    sol=Solution()
    List=head=ListNode(1)
    lst=[4,3,2,5,2]
    for x in lst:
        List.next=ListNode(x)
        List=List.next
    head1=sol.partition(head,3)
    print 'fcuk'
    while head1:
        print head1.val
        head1=head1.next

File: Algorithm/CPP/partition_list.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <string>
# include <unordered_map>
# include <cctype>
# include <algorithm>
# include <time.h>
using namespace std;

class Solution {
public:
	ListNode* partition(ListNode* head, int x) {
		auto dummy1 = new ListNode(-1);
		auto dummy2 = new ListNode(-1);
		auto p = dummy1;
		auto q = dummy2;
		while (head)
		{
			if (head->val <x )
			{
				p->next = head;
				p = p->next;
			}
			else
			{
				q->next = head;
				q = q->next;
			}
			head = head->next;
		}
		q->next = NULL;
		p->next = dummy2->next;
		return dummy1->next;

	}
};
int main()
{
	ListNode *head1 = new ListNode(1);
	ListNode *head2 = new ListNode(-2);
	auto p = head1;
	auto q = head2;
	vector<int> vec{ 4,3,2,5,2};//2
	for (auto x : vec)
	{
		p->next = new ListNode(x);
		p = p->next;
	}
	p->next = NULL;
	vec = { 3, 4, 7 };
	for (auto v : vec)
	{
		q->next = new ListNode(v);
		q = q->next;
	}
	q->next = NULL;
	//p = root->next;

	Solution sol;

	auto pp = sol.partition(head1, 3);
	while (pp != NULL)
	{
		cout << pp->val << endl;
		pp = pp->next;
	}
	cout << endl;
	system("pause");
	return 0;
}
